Mathlib/Order/WellQuasiOrder.lean

```diff
@@ -0,0 +1,151 @@
+    WellQuasiOrdered fun a b : α × β ↦ r a.1 b.1 ∧ s a.2 b.2 := by
+  intro f
+  obtain ⟨g, h₁⟩ := hr.exists_monotone_subseq (Prod.fst ∘ f)
+  obtain ⟨m, n, h, hf⟩ := hs (Prod.snd ∘ f ∘ g)
+  exact ⟨g m, g n, g.strictMono h, h₁ _ _ h.le, hf⟩
+
+/-- A typeclass for an ordered with a well quasi-ordered `≤` relation.
+
+Note that this is unlike `WellFoundedLT`, which instead takes a `<` relation. -/
+@[mk_iff wellQuasiOrderedLE_iff']
+class WellQuasiOrderedLE (α : Type*) [LE α] where
+  wqo : @WellQuasiOrdered α (· ≤ ·)
+
+theorem wellQuasiOrdered_le [LE α] [h : WellQuasiOrderedLE α] : @WellQuasiOrdered α (· ≤ ·) :=
+  h.wqo
+
+section Preorder
+variable [Preorder α]
+
+instance (priority := 100) Finite.to_wellQuasiOrderedLE [Finite α] : WellQuasiOrderedLE α where
+  wqo := Finite.wellQuasiOrdered _
+
+instance (priority := 100) WellQuasiOrderedLE.to_wellFoundedLT [WellQuasiOrderedLE α] :
+    WellFoundedLT α := by
+  rw [WellFoundedLT, isWellFounded_iff, RelEmbedding.wellFounded_iff_no_descending_seq]
+  refine ⟨fun f ↦ ?_⟩
+  obtain ⟨a, b, h, hf⟩ := wellQuasiOrdered_le f
+  exact (f.map_rel_iff.2 h).not_le hf
+
+theorem WellQuasiOrdered.wellFounded {α : Type*} {r : α → α → Prop} [IsPreorder α r]
+    (h : WellQuasiOrdered r) : WellFounded fun a b ↦ r a b ∧ ¬ r b a := by
+  letI : Preorder α :=
+    { le := r
+      le_refl := refl_of r
+      le_trans := fun _ _ _ => trans_of r }
+  have : WellQuasiOrderedLE α := ⟨h⟩
+  exact wellFounded_lt
+
+theorem WellQuasiOrderedLE.finite_of_isAntichain [WellQuasiOrderedLE α] {s : Set α}
+    (h : IsAntichain (· ≤ ·) s) : s.Finite :=
+  h.finite_of_wellQuasiOrdered wellQuasiOrdered_le
+
+/-- A preorder is well quasi-ordered iff it's well-founded and has no infinite antichains. -/
+theorem wellQuasiOrderedLE_iff :
+    WellQuasiOrderedLE α ↔ WellFoundedLT α ∧ ∀ s : Set α, IsAntichain (· ≤ ·) s → s.Finite := by
+  refine ⟨fun h ↦ ⟨h.to_wellFoundedLT, fun s ↦ h.finite_of_isAntichain⟩,
+    fun ⟨hwf, hc⟩ ↦ ⟨fun f ↦ ?_⟩⟩
+  obtain ⟨g, h1 | h2⟩ := exists_increasing_or_nonincreasing_subseq (· > ·) f
+  · exfalso
+    apply RelEmbedding.not_wellFounded_of_decreasing_seq _ hwf.wf
+    exact (RelEmbedding.ofMonotone _ h1).swap
+  · contrapose! hc
+    refine ⟨Set.range (f ∘ g), ?_, ?_⟩
+    · rintro _ ⟨m, rfl⟩ _ ⟨n, rfl⟩ _ hf
+      obtain h | rfl | h := lt_trichotomy m n
+      · exact hc _ _ (g.strictMono h) hf
+      · contradiction
+      · exact h2 _ _ h (lt_of_le_not_le hf (hc _ _ (g.strictMono h)))
+    · refine Set.infinite_range_of_injective fun m n (hf : f (g m) = f (g n)) ↦ ?_
+      obtain h | rfl | h := lt_trichotomy m n <;>
+        (first | rfl | cases (hf ▸ hc _ _ (g.strictMono h)) le_rfl)
+
+instance [WellQuasiOrderedLE α] [Preorder β] [WellQuasiOrderedLE β] : WellQuasiOrderedLE (α × β) :=
+  ⟨wellQuasiOrdered_le.prod wellQuasiOrdered_le⟩
+
+end Preorder
+
+section LinearOrder
+variable [LinearOrder α]
+
+/-- In a linear order, WQOs and well-orders are equivalent. -/
```

Comment:
```suggestion
/-- A linear WQO is the same thing as a well-order. -/
```

Reply:
Hmm, I'd prefer
```suggestion
/-- In a linear order, the properties of being a WQO and a well-orders are equivalent. -/
```
since the original doesn't exactly typecheck but the first suggestion can be misinterpreted as "alpha (is a linear order + wqo) iff (it is a well-order)" which isn't what the statement is saying

Reply:
> the first suggestion can be misinterpreted as "alpha (is a linear order + wqo) iff (it is a well-order)" which isn't what the statement is saying

Since well-order means `LinearOrder` + `WellFoundedLT`, it actually kind of is what it's saying

Reply:
Yes I agree it's what you're saying, but it's not what the theorem is saying!

Reply:
I prefer Yaël's suggestion. The only difference between that statement and the theorem I prove is that the `LinearOrder` assumption isn't duplicated, which to me seems like a triviality.